File: millapp/materiais/report/localizador_de_modelo/localizador_de_modelo.js
// Copyright (c) 2025, Felipe and contributors
// For license information, please see license.txt

frappe.query_reports["Localizador de Modelo"] = {
	filters: [
		{
			"fieldname": "artigo",
			"label": __("Artigo"),
			"fieldtype": "Link",
			"options": "Artigos",
			"reqd": 1,
			"on_change": function () {
				const artigo = frappe.query_report.get_filter_value("artigo");
				console.log(artigo);
				if (!artigo) {
					frappe.msgprint(__("Por favor, selecione um Artigo primeiro."));
					return;
				}
				const modelos_filter = frappe.query_report.get_filter("modelos");
				if (modelos_filter) {
					modelos_filter.get_data().then(data => {
						frappe.query_report.set_filter_value("modelos", data.map(modelo => modelo.value));
					});
				}
			}
		},
		{
			"fieldname": "modelos",
			"label": __("Modelos"),
			"fieldtype": "MultiSelectList",
			get_data: function (txt) {
				const artigo = frappe.query_report.get_filter_value("artigo");
				console.log(artigo);
				if (!artigo) {
					frappe.msgprint(__("Por favor, selecione um Artigo primeiro."));
					return [];
				}

				return frappe.call({
					method: "millapp.apis.utils.get_filhos",
					args: {
						doctype: "Modelos de Artigos",
						parent_name: artigo
					}
				}).then(r => {
					const modelos = (r.message || []).map(modelo => ({
						value: modelo.name,
						description: modelo.name
					}));

					frappe.query_report.set_filter_value("modelos", modelos.map(modelo => modelo.value));

					return modelos;
				});
			},
			"reqd": 1
		},
		{
			"fieldname": "estado_pedido",
			"label": __("Estado do Pedido"),
			"fieldtype": "MultiSelectList",
			"options": ["Separado", "Entregue"],
			"default": ["Separado", "Entregue"],
			"reqd": 1
		},
		{
			"fieldname": "data_maxima_acerto",
			"label": __("Acerto Até"),
			"fieldtype": "Date",
			"default": frappe.datetime.nowdate(),
		},
	],
};

File: millapp/materiais/report/localizador_de_modelo/localizador_de_modelo.py
import frappe
from frappe import _


def execute(filters: dict | None = None):
    columns = get_columns()
    data = get_data(filters)
    
    return columns, data


def get_columns() -> list[dict]:
    return [
        {
            "label": _("Artigo"),
            "fieldname": "artigo",
            "fieldtype": "Data",
            "width": 150,
        },
        {
			"label": _("Tamanho"),
			"fieldname": "tamanho",
			"fieldtype": "Data",
			"width": 150,
		},
        {
            "label": _("Pedido"),
            "fieldname": "pedido",
            "fieldtype": "link",
            "options": "Pedidos",
			"width": 150,
        },
        {
            "label": _("Data Acerto"),
            "fieldname": "data_acerto",
            "fieldtype": "Date",
			"width": 150,
        },
        {
            "label": _("Quantidade"),
            "fieldname": "quantidade",
            "fieldtype": "Int",
			"width": 150,
        },
        {
            "label": _("Estado Pedido"),
            "fieldname": "estado_pedido",
            "fieldtype": "Data",
			"width": 150,
        },
    ]

def get_data(filters) -> list[list]:
    pedido_state = filters.get("estado_pedido", [])
    modelos = filters.get("modelos", [])
    data_maxima_acerto = filters.get("data_maxima_acerto", None)

    # Verifica se os filtros estão vazios e ajusta
    if not pedido_state:
        pedido_state = ['Separado', 'Entregue']  # Defina um valor padrão se necessário
    if not modelos:
        modelos = ['default_modelo']  # Defina um valor padrão para modelos, caso não seja passado

    # Consulta SQL com parâmetros tratados
    query = """
        SELECT
            map.artigo,
            mda.tamanho,
			ped.name AS pedido,
            ped.data_acerto,
            map.quantidade_entregue as quantidade,
            ped.pedido_state as estado_pedido
        FROM 
            `tabModelos do Artigo no Pedido` AS map
        INNER JOIN 
            `tabPedidos` AS ped ON map.parent = ped.name
        LEFT JOIN
			`tabModelos de Artigos` AS mda ON map.modelo = mda.name
        WHERE 
            map.modelo IN %(modelos)s
            AND ped.pedido_state IN %(pedido_state)s
            AND (
                ped.pedido_state != 'Entregue' -- Para estados diferentes de "Entregue", sem considerar data_acerto
                OR (
                    ped.pedido_state = 'Entregue'
                    AND (
                        %(data_maxima_acerto)s IS NULL
                        OR ped.data_acerto <= %(data_maxima_acerto)s
                    )
                )
            );
    """
    # Passa os parâmetros para a consulta
    data = frappe.db.sql(query, {"modelos": modelos, "pedido_state": pedido_state, "data_maxima_acerto": data_maxima_acerto}, as_dict=True)
    
    return data
